### PR TITLE
Corrolate message directly with batch processing

### LINKED ISSUE #11494: Direct message correlation on same partition
**Description**

diagram_1

Message correlation is a complex execution topic, which might involve multiple partition, buffering messages and correlation messages to the right instances.

If we reach a message catch event in a process instance, like above, then we send a command (via side-effects) to open a subscription and later requests are send to ack the subscription and correlate a message to that process instance. This involves multiple requests and commands. If we have only one partition or or the message is stored on the same partition the sending of such requests are not really necessary, (probably the subscription isn't really necessary as well).

When reaching the catch event, we know what the subscription partition is, if this is the same as our current partition we could just write a follow up command to drive the execution further. Without sending the command over the wire to ourself.

This is quite interesting in the case we have batch processing enabled https://github.com/camunda/zeebe/issues/11416 since we can directly correlated the buffered message and execute the catch event, without stopping.

**Example**

If we run following test:

```java
  private static final BpmnModelInstance SINGLE_MESSAGE_PROCESS =
      Bpmn.createExecutableProcess(PROCESS_ID)
          .startEvent()
          .intermediateCatchEvent("receive-message")
          .message(m -> m.name("message").zeebeCorrelationKeyExpression("key"))
          .endEvent()
          .done();

  @Test
  public void shouldCorrelateMessageIfPublishedBefore() {
    // given
    engine.deployment().withXmlResource(SINGLE_MESSAGE_PROCESS).deploy();

    engine
        .message()
        .withName("message")
        .withCorrelationKey("order-123")
        .withVariables(asMsgPack("foo", "bar"))
        .publish();

    // when
    final long processInstanceKey =
        engine
            .processInstance()
            .ofBpmnProcessId(PROCESS_ID)
            .withVariable("key", "order-123")
            .create();

    // then
    final Record<ProcessInstanceRecordValue> event =
        RecordingExporter.processInstanceRecords()
            .withElementId("receive-message")
            .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED)
            .getFirst();
    final Map<String, String> variables =
        ProcessInstances.getCurrentVariables(processInstanceKey, event.getPosition());
    assertThat(variables).containsOnly(entry("key", "\"order-123\""), entry("foo", "\"bar\""));

    RecordLogger.logRecords();
  }
```

We can see that the complete PI is executed in one batch:

```
--------
	['C'ommand/'E'event/'R'ejection] [valueType] [intent] - #[position]->#[source record position] K[key] - [summary of value]
	P9K999 - key; #999 - record position; "ID" element/process id; @"elementid"/[P9K999] - element with ID and key
	Keys are decomposed into partition id and per partition key (e.g. 2251799813685253 -> P1K005). If single partition, the partition is omitted.
	Long IDs are shortened (e.g. 'startEvent_5d56488e-0570-416c-ba2d-36d2a3acea78' -> 'star..acea78'
--------
C DPLY         CREATE            - #01-> -1  -1 - 
E PROC         CREATED           - #02->#01 K01 - process.xml -> "process" (version:1)
E DPLY         CREATED           - #03->#01 K02 - process.xml
E DPLY         FULLY_DISTRIBUTED - #04->#01 K02 - 
C MSG          PUBLISH           - #05-> -1  -1 - "message" correlationKey: order-123 with variables: {foo=bar}
E MSG          PUBLISHED         - #06->#05 K03 - "message" correlationKey: order-123 with variables: {foo=bar}
C CREA         CREATE            - #07-> -1  -1 - new <process "process"> (default start)  with variables: {key=order-123}
E VAR          CREATED           - #08->#07 K05 - key->"order-123" in <process [K04]>
C PI           ACTIVATE          - #09->#07 K04 - PROCESS "process" in <process "process"[K04]>
E CREA         CREATED           - #10->#07 K06 - new <process "process"> (default start)  with variables: {key=order-123}
E PI           ACTIVATING        - #11->#07 K04 - PROCESS "process" in <process "process"[K04]>
E PI           ACTIVATED         - #12->#07 K04 - PROCESS "process" in <process "process"[K04]>
C PI           ACTIVATE          - #13->#07  -1 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
E PI           ACTIVATING        - #14->#07 K07 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
E PI           ACTIVATED         - #15->#07 K07 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
C PI           COMPLETE          - #16->#07 K07 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
E PI           COMPLETING        - #17->#07 K07 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
E PI           COMPLETED         - #18->#07 K07 - START_EVENT "startEv..babc08b" in <process "process"[K04]>
E PI           SEQ_FLOW_TAKEN    - #19->#07 K08 - SEQUENCE_FLOW "sequenc..37862e1" in <process "process"[K04]>
C PI           ACTIVATE          - #20->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
E PI           ACTIVATING        - #21->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
E PROC_MSG_SUB CREATING          - #22->#07 K10 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> (no vars)
C MSG_SUB      CREATE            - #23->#07  -1 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> with variables: {foo=bar}
E PI           ACTIVATED         - #24->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
E MSG_SUB      CREATED           - #25->#07 K11 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> (no vars)
E MSG_SUB      CORRELATING       - #26->#07 K11 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> with variables: {foo=bar}
C PROC_MSG_SUB CORRELATE         - #27->#07  -1 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> with variables: {foo=bar}
E PROC_MSG_SUB CORRELATED        - #28->#07 K10 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> with variables: {foo=bar}
E PROC_EVNT    TRIGGERING        - #29->#07 K12 -  @"receive-message"[K09] in <process K01[K04]> with variables: {foo=bar}
C PI           COMPLETE          - #30->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
C MSG_SUB      CORRELATE         - #31->#07  -1 - "message" (inter.) @[K09] in <process "process"[K04]> (no vars)
E PI           COMPLETING        - #32->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
E VAR          CREATED           - #33->#07 K13 - foo->"bar" in <process [K04]>
E PI           COMPLETED         - #34->#07 K09 - INTERMEDIATE_CATCH_EVENT "receive-message" in <process "process"[K04]>
E PI           SEQ_FLOW_TAKEN    - #35->#07 K14 - SEQUENCE_FLOW "sequenc..58d91da" in <process "process"[K04]>
C PI           ACTIVATE          - #36->#07 K15 - END_EVENT "endEven..15bea3e" in <process "process"[K04]>
E MSG_SUB      CORRELATED        - #37->#07 K11 - "message" (inter.) correlationKey: order-123 @[K09] in <process "process"[K04]> with variables: {foo=bar}
E PI           ACTIVATING        - #38->#07 K15 - END_EVENT "endEven..15bea3e" in <process "process"[K04]>
E PI           ACTIVATED         - #39->#07 K15 - END_EVENT "endEven..15bea3e" in <process "process"[K04]>
E PI           COMPLETING        - #40->#07 K15 - END_EVENT "endEven..15bea3e" in <process "process"[K04]>
E PI           COMPLETED         - #41->#07 K15 - END_EVENT "endEven..15bea3e" in <process "process"[K04]>
C PI           COMPLETE          - #42->#07 K04 - PROCESS "process" in <process "process"[K04]>
E PI           COMPLETING        - #43->#07 K04 - PROCESS "process" in <process "process"[K04]>
E PI           COMPLETED         - #44->#07 K04 - PROCESS "process" in <process "process"[K04]>
```

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

Based on https://github.com/camunda/zeebe/pull/11408 I cherry-picked the changes for correlating directly (via writing follow-up commands). I rebased the changes with the real batch processing https://github.com/camunda/zeebe/pull/11531

**Details to the PR:**

When message correlation is happening on the same partition, we can directly write a follow-up command instead of sending an inter partition request.

This PR adds some conditional checks in order to distinguish between correlating to a different partition or to the same partition.

In combination with batch processing, this means that we can also directly process the follow-up command in the same batch and continue with the process instance execution. This allows to remove of several commit latency wait times.

Before: this command has been sent out, over loopback it has been written to the log. It needed to be committed and read again by the processor in order to continue. Benchmarks show that it can help significantly on
certain user bases, which control the message distribution and correlations.

For example, Falko has some benchmark run for certain user scenarios, where we reduce the complete process instance execution time by another 1 sec. https://camunda.slack.com/archives/C049E6APFNZ/p1675754034646389 

We can use this PR to verify whether it works with our benchmarks and in certain scenarios 

## Related issues

closes  https://github.com/camunda/zeebe/issues/11494

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/message/command/SubscriptionCommandSenderTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.protocol.Protocol;
+import io.camunda.zeebe.stream.api.InterPartitionCommandSender;
+import io.camunda.zeebe.stream.api.ProcessingResultBuilder;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import org.agrona.DirectBuffer;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class SubscriptionCommandSenderTest {
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/message/command/SubscriptionCommandSenderTest.java:25]: I was not sure how to test this behavior now. Ideally, we could test if certain commands come in and specific results are generated by the Engine. Then we could verify whether the result contains follow-up commands or side-effects for the interpretation command. But this is not possible right now.

I felt this test might be close to the implementation/new behavior which might make sense. Let me know what you think and whether you dislike it. If you ok with it I can also add tests for the other methods.

**Comment by remcowesterhoud** [comment]
@Zelldon I just want to quickly verify I understand the changes.

image
This is the current situation. We write a bunch of inter-partition command (even if they are on the same partition). Every single one of these red-dotted lines means we cannot continue the batch and have to start a new one. Costing us performance.

image
This would be the situation after the change (only if the process instance partition and the message partition are the same). Because we no longer need to do any inter-partition command sending we can process all these commands in a single batch, increasing the performance.

Is this correct?

File: engine/src/test/java/io/camunda/zeebe/engine/processing/message/command/SubscriptionCommandSenderTest.java (hunk 1 above)
Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/message/command/SubscriptionCommandSenderTest.java:25]: I am fine with these tests. I think the behavior is already covered by the qa tests. Here we test message publication and correlation for both a single partition and a multi-partition cluster.

**Review by remcowesterhoud — COMMENTED** [review]
@Zelldon I've asked a question if my understanding of the problem is correct. If it is the changes look good to me 👍

**Review by lenaschoenburg — APPROVED** [review]

**Comment by Zelldon** [comment]
@remcowesterhoud thanks so much for this quick review ! Sorry I had to delay this so much, was out for a while and busy with other things :see_no_evil: 

Yes it is correct what you wrote here https://github.com/camunda/zeebe/pull/11549#issuecomment-1429515764 that was the idea.

Thanks all. Then I will extend the test class and start to merge it.

**Comment by Zelldon** [comment]
I have extended the tests :white_check_mark: 

bors r+

**Pull request merged**